File: blueluak-jse/src/main/java/net/blueva/luak/lib/jse/JavaArray.java
/******************************************************************************
 *  ____  _            _                _  __
 * | __ )| |_   _  ___| |   _   _  __ _| |/ /
 * |  _ \| | | | |/ _ \ |  | | | |/ _` | ' /
 * | |_) | | |_| |  __/ |__| |_| | (_| | . \
 * |____/|_|\__,_|\___|_____\__,_|\__,_|_|\_\
 *
 *  BlueLuaK
 *  https://github.com/BluevaDevelopment/BlueLuaK
 *
 *  Based on LuaJ (https://luaj.org)
 *  Original work Copyright (c) 2009 Luaj.org
 *  Modifications Copyright (c) 2026 Blueva Development
 *
 *  SPDX-License-Identifier: MIT
 ******************************************************************************/
package net.blueva.luak.lib.jse;

import java.lang.reflect.Array;

import net.blueva.luak.LuaTable;
import net.blueva.luak.LuaUserdata;
import net.blueva.luak.LuaValue;
import net.blueva.luak.lib.OneArgFunction;

/**
 * LuaValue that represents a Java instance of array type.
 * <p>
 * Can get elements by their integer key index, as well as the length.
 * <p>
 * This class is not used directly.  
 * It is returned by calls to {@link CoerceJavaToLua#coerce(Object)} 
 * when an array is supplied.
 * @see CoerceJavaToLua
 * @see CoerceLuaToJava
 */
class JavaArray extends LuaUserdata {

	private static final class LenFunction extends OneArgFunction {
		public LuaValue call(LuaValue u) {
			return LuaValue.valueOf(Array.getLength(((LuaUserdata)u).m_instance));
		}
	}

	static final LuaValue LENGTH = valueOf("length");
	
	static final LuaTable array_metatable;
	static {
		array_metatable = new LuaTable();
		array_metatable.rawset(LuaValue.LEN, new LenFunction());
	}
	
	JavaArray(Object instance) {
		super(instance);
		setmetatable(array_metatable);
	}
	
	public LuaValue get(LuaValue key) {
		if ( key.equals(LENGTH) )
			return valueOf(Array.getLength(m_instance));
		if ( key.isint() ) {
			int i = key.toint() - 1;
			return i>=0 && i<Array.getLength(m_instance)?
				CoerceJavaToLua.coerce(Array.get(m_instance,key.toint()-1)):
				NIL;
		}
		return super.get(key);
	}

	public void set(LuaValue key, LuaValue value) {
		if ( key.isint() ) {
			int i = key.toint() - 1;
			if ( i>=0 && i<Array.getLength(m_instance) )
				Array.set(m_instance,i,CoerceLuaToJava.coerce(value, m_instance.getClass().getComponentType()));
			else if ( m_metatable==null || ! settable(this,key,value) )
					error("array index out of bounds");
		}
		else
			super.set(key, value);
	} 	
}

File: luak-jvm/src/main/kotlin/org/basaltmc/luak/lib/jvm/JavaArray.kt
/******************************************************************************
 *  _                _
 * | |   _   _  __ _| | __
 * | |  | | | |/ _` | |/ /
 * | |__| |_| | (_| |   <
 * |_____\__,_|\__,_|_|\_\
 *
 *  Luak
 *  https://github.com/BasaltProject/Luak
 *
 *  Based on LuaJ (https://luaj.org)
 *  Original work Copyright (c) 2009 Luaj.org
 *  Modifications Copyright (c) 2026 Basalt Project
 *
 *  SPDX-License-Identifier: MIT
 ******************************************************************************/
package org.basaltmc.luak.lib.jvm

import org.basaltmc.luak.LuaTable
import org.basaltmc.luak.LuaUserdata
import org.basaltmc.luak.LuaValue
import org.basaltmc.luak.lib.OneArgFunction
import java.lang.reflect.Array

/**
 * LuaValue that represents a Java instance of array type.
 * 
 * 
 * Can get elements by their integer key index, as well as the length.
 * 
 * 
 * This class is not used directly.
 * It is returned by calls to [CoerceJavaToLua.coerce]
 * when an array is supplied.
 * @see CoerceJavaToLua
 * 
 * @see CoerceLuaToJava
 */
internal class JavaArray(instance: Any?) : LuaUserdata(instance!!) {
    private class LenFunction : OneArgFunction() {
        override fun call(u: LuaValue?): LuaValue? {
            return valueOf(Array.getLength((u!! as LuaUserdata).m_instance))
        }
    }

    init {
        setmetatable(array_metatable)
    }

    override fun get(key: LuaValue): LuaValue {
        if (key == LENGTH) return valueOf(Array.getLength(m_instance))
        if (key!!.isint()) {
            val i = key.toint() - 1
            return if (i >= 0 && i < Array.getLength(m_instance)) CoerceJavaToLua.coerce(
                Array.get(
                    m_instance,
                    key.toint() - 1
                )
            )!! else NIL!!
        }
        return super.get(key)
    }

    override fun set(key: LuaValue?, value: LuaValue?) {
        val key = key!!
        if (key.isint()) {
            val i = key.toint() - 1
            if (i >= 0 && i < Array.getLength(m_instance)) Array.set(
                m_instance,
                i,
                CoerceLuaToJava.coerce(value, m_instance.javaClass.getComponentType())
            )
            else if (m_metatable == null || !settable(this, key, value)) error("array index out of bounds")
        } else super.set(key, value)
    }

    companion object {
        val LENGTH: LuaValue = valueOf("length")

        val array_metatable: LuaTable

        init {
            array_metatable = LuaTable()
            array_metatable.rawset(LEN, LenFunction())
        }
    }
}
